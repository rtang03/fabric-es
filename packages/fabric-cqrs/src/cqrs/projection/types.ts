import { Commit } from '../../types';

export interface FindAction {
  type: string;
  payload?: {
    tx_id: string;
    args: {
      where?: Record<string, any>;
      all?: boolean;
      contain?: string | number;
    };
  };
}

export interface UpsertAction {
  type: string;
  payload?: {
    tx_id: string;
    args: {
      commit: Record<string, Commit>;
    };
  };
}

export interface UpsertManyAction {
  type: string;
  payload?: {
    tx_id: string;
    args: {
      commits: Record<string, Commit>;
    };
  };
}
