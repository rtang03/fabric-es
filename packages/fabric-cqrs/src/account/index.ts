// export * from './registerUser';
export * from './enrolAdmin';
