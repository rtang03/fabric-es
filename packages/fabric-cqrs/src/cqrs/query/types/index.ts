/**
 * @packageDocumentation
 * @hidden
 */
export * from './action';
