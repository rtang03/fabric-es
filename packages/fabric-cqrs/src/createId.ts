export const createId: (parts: string[]) => string = parts => parts[0] + '::' + parts[1];
