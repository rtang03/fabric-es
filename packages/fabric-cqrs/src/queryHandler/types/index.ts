export * from './commitHashfields';
export * from './entityDefaultSchema';
export * from './commitDefaultSchema';
//
export * from './mapField';
export * from './redisCommit';
