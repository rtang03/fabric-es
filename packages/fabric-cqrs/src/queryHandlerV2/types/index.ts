export * from './mapField';
export * from './redisCommitFields';
export * from './redisRepository';
export * from './queryDatabaseResponse';
export * from './queryDatabaseV2';
export * from './reselectedCommit';
