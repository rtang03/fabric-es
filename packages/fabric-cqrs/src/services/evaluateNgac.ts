import { Network } from 'fabric-network';
import { getContract } from './contract';

export const evaluateNgac: <T = any>(
  fcn: string,
  args: string[],
  { network }: {network: Network}
) => Promise<T | { error: any }> = async <T = any>(fcn, args, { network }) =>
  await getContract(network).then(
    async ({ contract }) =>
      await contract
        .createTransaction(fcn)
        .evaluate(...args)
        .then<T>((res: any) =>
          JSON.parse(Buffer.from(JSON.parse(res)).toString())
        )
        .catch(error => {
          console.log(`Error processing Evaluate transaction: ${fcn}`);
          // console.error(error.stack);
          return { error };
        })
  );

export default evaluateNgac;
