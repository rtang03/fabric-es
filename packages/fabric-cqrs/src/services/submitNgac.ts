import { Context } from '../types';
import { getContract } from './contract';

export const submitNgac: <T = any>(
  fcn: string,
  args: string[],
  { network }: Context
) => Promise<T | { error?: any; status?: string; message?: string }> = async <
  T = any
>(
  fcn,
  args,
  { network }
) =>
  await getContract(network).then(({ contract }) =>
    contract
      .createTransaction(fcn)
      .submit(...args)
      .then<T>((res: any) =>
        JSON.parse(Buffer.from(JSON.parse(res)).toString())
      )
      .catch(error => {
        console.log(`Error processing Submit transaction.`);
        console.error(error.stack);
        console.log(fcn);
        return { error };
      })
  );

export default submitNgac;
