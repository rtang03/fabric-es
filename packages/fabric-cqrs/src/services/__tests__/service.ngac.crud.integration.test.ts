import { Gateway, Network } from 'fabric-network';
import { bootstrap } from '../../account';
import { createId } from '../../createId';
import { Policy } from '../../types';
import evaluateNgac from '../evaluateNgac';
import submitNgac from '../submitNgac';

let network: Network;
let gateway: Gateway;

const identity = `service_ngac_test${Math.floor(Math.random() * 1000)}`;
const id = createId(['Org1MSP', 'Admin@org1.example.com']);
const entityName = 'ngac_service_test';
const entityId = 'entity_id_1001';

beforeAll(async () => {
  const config = await bootstrap(identity);
  network = config.network;
  gateway = config.gateway;
});

afterAll(async () => await gateway.disconnect());

describe('Ngac CRUD Integration Tests', () => {
  it('should getMSPAttrByMSPID', async () =>
    evaluateNgac('getMSPAttrByMSPID', ['Org1MSP'], {
      network
    }).then(attributes =>
      expect(attributes).toEqual([
        { type: '1', key: 'mspid', value: 'Org1MSP' }
      ])
    ));

  it('should getResourceAttrByURI', async () =>
    evaluateNgac('getResourceAttrByURI', ['model/Org1MSP/ngactest'], {
      network
    }).then(attributes => expect(attributes).toMatchSnapshot()));

  it('should getPolicyById', async () =>
    evaluateNgac('getPolicyById', [id], { network }).then(
      (policies: Policy[]) =>
        policies.forEach(({ key }) =>
          expect(key.startsWith(JSON.stringify(id))).toBe(true)
        )
    ));

  it('should getPolicyByIdSid', async () =>
    evaluateNgac('getPolicyByIdSid', [id, 'allowCreateTest'], { network }).then(
      policy => expect(policy).toMatchSnapshot()
    ));

  it('should addPolicy', async () => {
    const pClass = 'service-ngac-crud-policy';
    const sid = 'allowSvsCreate';
    const url = 'model/Org1MSP/service_ngac_crud_test/crud_int_001';
    const events = ['ServiceNgacCRUDTestCreated'];
    await submitNgac(
      'addPolicy',
      [pClass, sid, url, JSON.stringify(events), ''],
      {
        network
      }
    ).then(policy => expect(policy.sid).toEqual(sid));
  });

  it('should deletePolicyByIdSid', async () =>
    submitNgac('deletePolicyByIdSid', [id, 'allowSvsCreate'], {
      network
    }).then(({ status }) => expect(status).toEqual('SUCCESS')));

  it('should addMSPAttr', async () => {
    const mspAttr = [{ type: '1', key: 'mspid', value: 'Org100MSP' }];
    await submitNgac('addMSPAttr', ['Org100MSP', JSON.stringify(mspAttr)], {
      network
    }).then(attributes => expect(attributes).toEqual(mspAttr));
  });

  it('should deleteMSPAttrByMSPID', async () =>
    submitNgac('deleteMSPAttrByMSPID', ['Org100MSP'], { network }).then(
      result =>
        expect(result).toEqual({
          status: 'SUCCESS',
          message: 'Org100MSP is successfully deleted'
        })
    ));

  it('should addResourceAttr', async () => {
    // note: this api use submitter's mspId, instead of input argument
    // Todo: may require to visit if this behaviour is correct.
    const resourceAttr = [{ type: '1', key: 'creatTest', value: 'any value' }];
    await submitNgac(
      'addResourceAttr',
      [entityName, entityId, JSON.stringify(resourceAttr)],
      {
        network
      }
    ).then(attributes => expect(attributes).toEqual(resourceAttr));
    await evaluateNgac(
      'getResourceAttrByURI',
      [`model/Org1MSP/${entityName}/${entityId}`],
      {
        network
      }
    ).then(attributes => expect(attributes).toEqual(resourceAttr));
  });

  it('should upsertResourceAttr', async () => {
    const resourceAttr = [{ type: '1', key: 'username', value: 'bob' }];
    const expectedResult = [
      { type: '1', key: 'creatTest', value: 'any value' },
      { type: '1', key: 'username', value: 'bob' }
    ];
    await submitNgac(
      'upsertResourceAttr',
      [entityName, entityId, JSON.stringify(resourceAttr)],
      { network }
    ).then(attributes => expect(attributes).toEqual(expectedResult));
    await evaluateNgac(
      'getResourceAttrByURI',
      [`model/Org1MSP/${entityName}/${entityId}`],
      {
        network
      }
    ).then(attributes => expect(attributes).toEqual(expectedResult));
  });

  it('should deleteReourceAttrByURI', async () => {
    await submitNgac(
      'deleteReourceAttrByURI',
      [`model/Org1MSP/${entityName}/${entityId}`],
      { network }
    ).then(({ status }) => expect(status).toEqual('SUCCESS'));
    await evaluateNgac(
      'getResourceAttrByURI',
      [`model/Org1MSP/${entityName}/${entityId}`],
      {
        network
      }
    ).then(attributes => expect(attributes).toEqual({}));
  });
});
