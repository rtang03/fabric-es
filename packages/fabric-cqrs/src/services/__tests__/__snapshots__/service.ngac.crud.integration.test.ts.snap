// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ngac CRUD Integration Tests should getPolicyByIdSid 1`] = `
Object {
  "allowedEvents": Array [
    "TestCreated",
  ],
  "attributes": Object {
    "uri": "model/Org1MSP/test/test_0001",
  },
  "effect": "Allow",
  "key": "\\"Org1MSP::Admin@org1.example.com\\"~\\"allowCreateTest\\"",
  "policyClass": "int-test",
  "sid": "allowCreateTest",
}
`;

exports[`Ngac CRUD Integration Tests should getResourceAttrByURI 1`] = `
Array [
  Object {
    "key": "createTest",
    "type": "1",
    "value": "\\"Org1MSP\\"~\\"Admin@org1.example.com\\"",
  },
]
`;
